README.md
<a href="https://repology.org/metapackage/libsdl2pp/versions">
    <img src="https://repology.org/badge/vertical-allrepos/libsdl2pp.svg" alt="libsdl2pp packaging status" align="right">
</a>

[![CI](https://github.com/libSDL2pp/libSDL2pp/actions/workflows/ci.yml/badge.svg)](https://github.com/libSDL2pp/libSDL2pp/actions/workflows/ci.yml)
[![Github commits (since latest release)](https://img.shields.io/github/commits-since/libSDL2pp/libSDL2pp/latest.svg)](https://github.com/libSDL2pp/libSDL2pp)

# libSDL2pp

This library provides C++ bindings/wrapper for SDL2 and satellite libraries.

## Synopsis

```cpp
try {
  using namespace SDL2pp;

  // Init SDL; will be automatically deinitialized when the object is destroyed
  SDL sdl(SDL_INIT_VIDEO | SDL_INIT_AUDIO);

  // Likewise, init SDL_ttf library
  SDLTTF sdl_ttf;

  // Straightforward wrappers around corresponding SDL2 objects
  // These take full care of proper object destruction and error checking
  Window window("libSDL2pp demo",
                SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED,
                640, 480, SDL_WINDOW_RESIZABLE);
  Renderer renderer(window, -1, SDL_RENDERER_ACCELERATED);
  Texture sprite1(renderer, SDL_PIXELFORMAT_ARGB8888,
                            SDL_TEXTUREACCESS_STATIC, 16, 16);
  Texture sprite2(renderer, "sprite.png"); // SDL_image support

  Font font("Vera.ttf", 20); // SDL_ttf font

  // Initialize audio mixer
  Mixer mixer(MIX_DEFAULT_FREQUENCY, MIX_DEFAULT_FORMAT, MIX_DEFAULT_CHANNELS, 4096);

  Chunk sound("effect.ogg"); // OGG sound file

  // Create texture from surface containing text rendered by SDL_ttf
  Texture text(renderer, font.RenderText_Solid("Hello, world!",
               SDL_Color{255, 255, 255, 255}));

  unsigned char pixels[16 * 16 * 4];

  // Note proper constructor for Rect
  sprite1.Update(Rect(0, 0, 16, 16), pixels, 16 * 4);

  // Most setter methods are chainable
  renderer.SetLogicalSize(640, 480).SetDrawColor(0, 16, 32).Clear();

  // Also note a safe way to specify null rects and points
  renderer.Copy(sprite1, NullOpt, NullOpt);

  // There are multiple convenient ways to construct e.g. a Rect;
  // Objects provide extensive set of getters
  renderer.Copy(text, NullOpt, Rect(Point(0, 0), text.GetSize()));

  // Copy() is overloaded, providing access to both SDL_RenderCopy and SDL_RenderCopyEx
  renderer.Copy(sprite2, NullOpt, NullOpt, 45.0);

  renderer.Present();

  // Play our sound one time on a first available mixer channel
  mixer.PlayChannel(-1, sound);

  // You can still access wrapped C SDL types
  SDL_Renderer* sdl_renderer = renderer.Get();

  // Of course, C SDL2 API is still perfectly valid
  SDL_Delay(2000);

  // All SDL objects are released at this point or if an error occurs
} catch (SDL2pp::Exception& e) {
  // Exception stores SDL_GetError() result and name of function which failed
  std::cerr << "Error in: " << e.GetSDLFunction() << std::endl;
  std::cerr << "  Reason: " << e.GetSDLError() << std::endl;
} catch (std::exception& e) {
  // This also works (e.g. "SDL_Init failed: No available video device")
  std::cerr << e.what() << std::endl;
}
```

There's also more elaborate [tutorial](https://github.com/libSDL2pp/libSDL2pp-tutorial).

## Features

Currently, the library provides wrapper classes for

* SDL
  * Library initialization/deinitialization
  * Audio
    * Audio device
    * WAV-related functions
    * SDL_AudioSpec
  * Graphics
    * SDL_Point
    * SDL_Rect
    * SDL_Renderer
    * SDL_Surface
    * SDL_Texture
    * SDL_Window
  * I/O
    * SDL_RWops
* SDL_image
  * Library initialization/deinitialization
  * Loading functions integrated into Texture and Surface
* SDL_mixer
  * Library initialization/deinitialization
  * Sound sample handling (Mix_Chunk)
  * Music handling (Mix_Music)
  * Mixer class which encapsulates device handling and all playback functions
* SDL_ttf
  * Library initialization/deinitialization
  * TTF_Font (all functions covered)

each with a subset of methods corresponding to SDL functions working
with specific types of objects and, in some cases, additional convenience
methods. These classes support:

* RAII-style initialization and destruction, automatic resource lifetime
  control (you no longer need to care of manually freeing your stuff)
* Total error checking: exceptions are thrown if any SDL function fails.
  Exception itself allows retrieval of SDL error string (you no longer
  need to manually check return code after each function call)
* Method overloading, default arguments, method chaining allow shorter
  and cleaner code
* Move semantics support, which allow you to store SDL objects
  in containers and pass/return them by value without noticeable overhead

Set of functional extensions above SDL2 is also available:

* RWops adapters for C++ containers and streams
* Optional object to safely handle values which may not be present,
  (for which SDL2 usually uses NULL pointers)
* Number of additional methods and operator support for Point and Rect

## Building

Dependencies:
* [cmake](https://cmake.org/)
* [SDL2](https://libsdl.org/)
* [SDL2_image](https://www.libsdl.org/projects/SDL_image/) (optional)
* [SDL2_mixer](https://www.libsdl.org/projects/SDL_mixer/) (optional)
* [SDL2_ttf](https://www.libsdl.org/projects/SDL_ttf/) (optional)

To build standalone version:

```shell
cmake .
cmake --build .
```

Following variables may be supplied to CMake to affect build:

* `SDL2PP_WITH_IMAGE` - enable SDL_image support (default ON)
* `SDL2PP_WITH_MIXER` - enable SDL_mixer support (default ON)
* `SDL2PP_WITH_TTF` - enable SDL_ttf support (default ON)
* `SDL2PP_WITH_EXAMPLES` - enable building example programs (only for standalone build, default ON)
* `SDL2PP_WITH_TESTS` - enable building tests (only for standalone build, default ON)
* `SDL2PP_STATIC` - build static library instead of shared (only for standalone build, default OFF)
* `SDL2PP_ENABLE_LIVE_TESTS` - enable tests which require X11 and/or audio device to run (only for standalone build, default ON)

## Installation

To install the library system-wide, run:

```shell
cmake .
cmake --build .
cmake --install .
```

You can change installation prefix with CMAKE_INSTALL_PREFIX cmake
variable:

```shell
cmake -DCMAKE_INSTALL_PREFIX=/usr/local .
cmake --build .
cmake --install .
```

SDL2pp installs `pkg-config` file, so it can be used with any build
system that interacts with `pkg-config`, including CMake, meson and
GNU Autotools. It also installs CMake module file, which can be used
from CMake directly:

```cmake
find_package(SDL2pp REQUIRED)

target_link_libraries(mytarget SDL2pp::SDL2pp)
```

## Bundling ##

The library is easy to integrate into other CMake projects
(and as the library has no stable API yet, this way of using it is
still recommended).

Just place the library into dedicated directory in your project
(for example, extlib/libSDL2pp) and add

```cmake
set(SDL2PP_WITH_IMAGE ON) # if you need SDL_image support
set(SDL2PP_WITH_MIXER ON) # if you need SDL_mixer support
set(SDL2PP_WITH_TTF ON) # if you need SDL_ttf support
add_subdirectory(extlib/libSDL2pp)
```

into your core CMakeLists.txt. This will act similar to how
`find_package` usually does, and will provide `SDL2pp::SDL2pp`
target for your project. You will then be able it as usual:

```cmake
target_link_libraries(mytarget SDL2pp::SDL2pp)
```

If bundled, libSDL2pp does not build examples and becomes a static
library. See [hoverboard](https://github.com/AMDmi3/hoverboard-sdl/blob/5729f9fb6929f6e8147481f5b21772fc1e35562a/CMakeLists.txt#L34-L40)
project as an example of using both bundled and systemwide SDL2pp.

## Completeness

The library still doesn't cover all aspects of SDL2, and the development
is generally guided by the author's needs and interest without a goal
for covering all SDL2 functions as soon as possible. However, if
you need specific bits which are not yet implemented in the library,
feel free to drop an issue. Patches are of course more than welcome.

It should be noted, however, that I currently do not plan to implement
any wrappers over non object-oriented SDL2 code, as these will not bring
any benefits over using plain C API. E.g. I see no point in implementing
SDL2pp::Delay() as it won't bring any convenience over SDL_Delay().

The same strongly applies to the SDL2 bits which duplicate C++17
standard library, e.g. threads and atomic ops.

## Users

Projects using libSDL2pp:

* [fontbm](https://github.com/vladimirgamalian/fontbm) - Command line font generator, compatible with BMFont
* [hoverboard-sdl](https://github.com/AMDmi3/hoverboard-sdl) - Desktop version of xkcd 1608 "Hoverboard" game
* [neopedersia](https://github.com/vladimirgamalian/neopedersia) - Nexus Wars style game
* [OpenDaed](https://github.com/AMDmi3/opendaed) - Libre reimplementation of The Daedalus Encounter game
* [OpenStrike](https://github.com/AMDmi3/openstrike) - Libre reimplementation of Jungle and Desert Strike games
* [osmview](https://bitbucket.org/ipopov/osmview) - Desktop OpenStreetMap viewer
* [planetonomy](https://github.com/AMDmi3/planetonomy) - Old-school platformer/exploration game with CGA graphics

## WWW

* [GitHub page](https://github.com/libSDL2pp/libSDL2pp)
* [Online documentation](https://sdl2pp.amdmi3.ru/)

## Author

* [Dmitry Marakasov](https://github.com/AMDmi3) <amdmi3@amdmi3.ru>

## Contributors

* [Aargonian](https://github.com/Aargonian)
* [Carl Schwope](https://github.com/Lowest0ne)
* [Carsten Elton Sorensen](https://github.com/csoren)
* [Ilya Popov](https://github.com/ilyapopov)
* [kumar8600](https://github.com/kumar8600)
* [ooxi](https://github.com/ooxi)
* [Vladimir Gamalian](https://github.com/vladimirgamalian)
* [Vraiment](https://github.com/Vraiment)

## License

libSDL2pp comes under zlib license, the same license as SDL2. See COPYING.txt.


CHANGES.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 0.18.1 - 2023-04-17
### Fixed
* Fix SDL2main library link order

## 0.18.0 - 2023-03-27
### Changed
* Switch SDL2_{mixer,mixer,ttf} detection to upstream CMake scripts

## 0.17.0 - 2022-04-28
### Changed
* Switch to modern CMake
* Switch to C++17

## 0.16.2 - 2022-04-25
### Fixed
* Fix font dimensions test

## 0.16.1 - 2020-06-18
### Fixed
* Make `SDLmain` library optional in Unix, since some distros seem to drop it
* Fix using unsupported warning flags on some compilers

## 0.16.0 - 2018-02-11
### Added
* New SDL 2.0.5 `Window` method: `Window::SetResizable()`
* Minor documentation improvements
* Example code fixes

## 0.15.0 - 2017-07-10
### Added
* `Color` class wrapping around `SDL_Color`

## 0.14.1 - 2017-06-28
### Fixed
* `Surface.hh` include was missing from `SDL2pp.hh`

## 0.14.0 - 2017-05-29
### Fixed
* Proper `Size()` handling for custom `RWops`

### Added
* `Window::GetOpacity()` and `Window()::SetOpacity()` wrappers for functions appeared in SDL 2.0.5
* `Music` RWops constructors based on undocumented SDL_mixer functions
* Static standalone build support

### Changed
* libSDL2pp now follows SDL2 include path conventions, finding and using SDL2 headers without SDL2/ prefix

## 0.13.0 - 2016-11-08
### Fixed
* Fixed SDL 2.0.4 specific `AudioDevice` methods: `AudioDevice::QueueAudio()`, `AudioDevice::GetQueuedAudioSize()`
* Fixed MSVC compiler support and .dll generation
* Multiple documentation fixes

### Added
* More tests
* Automatic SDL 2.0.4 detection
* `Texture::Update()` overload which takes rvalue reference to `Surface`

### Changed
* `Renderer::GetClipRect` now returns `Optional<Rect>` instead of (possibly empty) `Rect`
* `Exception` is now derived from `std::runtime_error`

### Removed
* Deprecated `Renderer::GetInfo()` variant which takes pointer (use variant which takes reference)

## 0.12.0 - 2016-01-23
### Fixed
* Link SDL2main library, which is needed on some systems
* Numerous improvements in Windows support
* Documentation fixes
* Multiple fixes in pkg-config and CMake module files

## 0.11.0 - 2015-11-30
### Added
* New `Window` methods: `Window::GetDrawableWidth()`, `Window::GetDrawableHeight()`, `Window::GetDrawableSize()` (@csoren)
* `Point` and `Rect` less-than operators and std::hash support which makes them usable as `std::map`/`std::unordered_map`/`std::set`/`std::unordered_set` keys
* More `Rect` methods and wider constexpr support

### Fixed
* `Rect` documentation fixes
* Unused variable warnings (@vladimirgamalian)
* Incorrect return value handling in `Surface::SetClipRect()`

## 0.10.0 - 2015-10-14
### Fixed
* Improved cygwin compatibility
* Missing `SDL` method implementations
* Fixed return and argument types for some `Window`, `Font` and `Mixer` methods to match upstream API
* Infinite loop in `StreamRWops`
* Documentation improvements
* Build without [SDL_mixer](https://www.libsdl.org/projects/SDL_mixer/) support

## 0.9.0 - 2015-09-07
### Added
* [SDL_mixer](https://www.libsdl.org/projects/SDL_mixer/) support
* More `Point` operators: `*`, `/`, `%` for integer multiplication, division and remainder by a number or another `Point`
* `Point` and `Rect` better inlining and constexpr support
* New method: `Surface::GetFormat()`
* `Texture::Update()` overload which takes Surface as pixel source

## 0.8.2 - 2015-07-28
### Fixed
* Add missing `Font::IsGlyphProvided()` implementation (@vladimirgamalian)
* Fix glyph rectangle calculation in `Font::GetGlyphRect()` (@vladimirgamalian)

## 0.8.1 - 2015-06-29
### Fixed
* Use correct type flags in `RWops::Size()`, fixing MinGW build (@vladimirgamalian)

## 0.8.0 - 2015-06-10
### Added
* `Render::Copy()` overload which takes target `Point` instead of a `Rect`
* New method `Render::FillCopy()` which fills target rect with repeated texture
* Improved UTF-16 font rendering support
* CMake module file

## 0.7.1 - 2015-03-27
### Fixed
* Miscellaneous style, documentation and build system fixes

## 0.7.0 - 2015-01-22
### Added
* More `Texture`, `Renderer` (both complete now) and `Window` methods
* New methods: `Point::Clamp()`, `Point::Wrap()`, `Rect::Extend()`, `Rect::IntersectLine()`
* STL stream output for `Point` and `Rect` (as in `std::cout << Rect(10, 20, 30 40);`)

### Changed
* Most setters now return reference to self to allow method chaining
* Exception now stores both SDL error and name of function which caused it; `what()` message now includes these both in a readable way, so SDL2pp exceptions may be conveniently handled with `catch (std::exception&)`
* pkg-config file now provides SDL libraries along with SDL2pp, so users don't need to care of SDL2 flags in addition to SDL2pp's

### Removed
* Deprecated `Point` and `Rect` methods: `Null()`, `IsNull()`, `Get()`

### Fixed
* Use move, not copy in constructing `RWops` from `CustomRWops`-derived class

## 0.6.0 - 2015-01-11
### Added
* [Doxygen](http://doxygen.org/) documentation for the whole library
* Wrapper around [SDL_image](https://www.libsdl.org/projects/SDL_image/) initialization/deinitialization: `SDLImage` class
* `Surface``, a wrapepr for [SDL_Surface](https://wiki.libsdl.org/SDL_Surface) including [SDL_image](https://www.libsdl.org/projects/SDL_image/) support. It's now also possible to construct `Texture` from `Surface`.
* Bunch of constructors from existing SDL objects (e.g. `Window::Window(SDL_Window*)`) for `Point`, `Rect`, `Window`, `Renderer`, `Texture`
* Complete [SDL_ttf](https://www.libsdl.org/projects/SDL_ttf/) support
* Consistent `GetWidth()`/`GetHeight()`/`GetSize()` for `Texture` and `Surface`
* More operators, constructors and methods for `Point` and `Rect`

### Fixed
* Added proper error checking for SDL_image-using Texture constructors
* Multiple fixes in build system

## 0.5.2 - 2014-12-30
### Fixed
* Added missing include file to installation

## 0.5.1 - 2014-12-27
### Fixed

demo/demo.cc
/*
  libSDL2pp - C++ wrapper for libSDL2
  Copyright (C) 2013 Dmitry Marakasov <amdmi3@amdmi3.ru>

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.
*/

#include <iostream>

#include <SDL2/SDL.h>

#include <SDL2pp/SDL2pp.hh>

using namespace SDL2pp;

#define RGBA(r, g, b, a) r, g, b, a
unsigned char pixels[4 * 4 * 4] = {
	RGBA(0xff, 0x00, 0x00, 0xff), RGBA(0xff, 0x80, 0x00, 0xff), RGBA(0xff, 0xff, 0x00, 0xff), RGBA(0x80, 0xff, 0x00, 0xff),
	RGBA(0xff, 0x00, 0x80, 0xff), RGBA(0xff, 0xff, 0xff, 0xff), RGBA(0x00, 0x00, 0x00, 0x00), RGBA(0x00, 0xff, 0x00, 0xff),
	RGBA(0xff, 0x00, 0xff, 0xff), RGBA(0x00, 0x00, 0x00, 0x00), RGBA(0x00, 0x00, 0x00, 0xff), RGBA(0x00, 0xff, 0x80, 0xff),
	RGBA(0x80, 0x00, 0xff, 0xff), RGBA(0x00, 0x00, 0xff, 0xff), RGBA(0x00, 0x80, 0xff, 0xff), RGBA(0x00, 0xff, 0xff, 0xff),
};

int Run() {
	SDL sdl(SDL_INIT_VIDEO);
	Window window("libSDL2pp demo", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, 640, 480, SDL_WINDOW_RESIZABLE);
	Renderer render(window, -1, SDL_RENDERER_ACCELERATED);
	Texture sprite(render, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_STATIC, 4, 4);

	sprite.Update(Rect::Null(), pixels, 4 * 4);
	sprite.SetBlendMode(SDL_BLENDMODE_BLEND);

	render.SetDrawBlendMode(SDL_BLENDMODE_BLEND);

	while (1) {
		// Process events
		SDL_Event event;
		while (SDL_PollEvent(&event)) {
			if (event.type == SDL_QUIT || (event.type == SDL_KEYDOWN && event.key.keysym.sym == SDLK_ESCAPE))
				return 0;
		}

		// Render
		render.SetDrawColor(0, 0, 0);
		render.Clear();

		// Render lines
		render.SetDrawColor(255, 0, 0);
		render.DrawLine(10, 10, 630, 10);
		render.SetDrawColor(0, 255, 0);
		render.DrawLine(630, 10, 630, 470);
		render.SetDrawColor(0, 0, 255);
		render.DrawLine(630, 470, 10, 470);
		render.SetDrawColor(255, 255, 255);
		render.DrawLine(10, 470, 10, 10);

		render.SetDrawColor(255, 255, 255, 127);
		render.FillRect(0, 0, 20, 20);
		render.SetDrawColor(255, 255, 255);
		render.DrawRect(0, 0, 20, 20);

		// Pixel-perfectness test
		render.SetDrawColor(192, 192, 192);
		render.DrawLine(6, 2, 6, 10);
		render.DrawLine(2, 6, 10, 6);

		render.SetDrawColor(255, 255, 255);
		render.DrawRect(5, 5, 7, 7);
		render.DrawRect(3, 3, 9, 9);

		// Render 4 smaller squares
		sprite.SetAlphaMod(0xff);
		render.Copy(sprite, Rect::Null(), Rect(80, 0, 240, 240), SDL_GetTicks() / 5000.0 * 360.0, Point::Null(), SDL_FLIP_NONE);
		render.Copy(sprite, Rect::Null(), Rect(80, 360, 240, 240), -1.0 * SDL_GetTicks() / 5000.0 * 360.0, Point::Null(), SDL_FLIP_NONE);
		render.Copy(sprite, Rect::Null(), Rect(400, 0, 240, 240), -1.0 * SDL_GetTicks() / 5000.0 * 360.0, Point::Null(), SDL_FLIP_NONE);
		render.Copy(sprite, Rect::Null(), Rect(400, 360, 240, 240), SDL_GetTicks() / 5000.0 * 360.0, Point::Null(), SDL_FLIP_NONE);

		// Render transparent bigger square
		sprite.SetAlphaMod(0x80);
		render.Copy(sprite, Rect::Null(), Rect(80, 0, 480, 480), SDL_GetTicks() / 10000.0 * 360.0, Point::Null(), SDL_FLIP_NONE);

		render.Present();

		// Frame limiter
		SDL_Delay(1);
	}

	return 0;
}

int main() {
	try {
		return Run();
	} catch (Exception& e) {
		std::cerr << "Error: " << e.what() << " (" << e.GetSDLError() << ")" << std::endl;
	} catch (std::exception& e) {
		std::cerr << "Error: " << e.what() << std::endl;
	}

	return -1;
}
